handheld-daemon: 3.6.1 -> 3.7.0

GaetanLepage opened the pull request:
## Things done

Changelog: https://github.com/hhd-dev/hhd/releases/tag/v3.7.0

cc @appsforartists @toast003

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

GaetanLepage committed:
handheld-daemon: 3.6.1 -> 3.6.2

Diff: https://github.com/hhd-dev/hhd/compare/refs/tags/v3.6.1...v3.6.2

Changelog: https://github.com/hhd-dev/hhd/releases/tag/v3.6.2

GaetanLepage commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 361364`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>handheld-daemon</li>
    <li>handheld-daemon.dist</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>handheld-daemon</li>
    <li>handheld-daemon.dist</li>
  </ul>
</details>

---
### `x86_64-darwin`

---
### `aarch64-darwin`

toast003 commented:
Looks good, the new hhdctl command works

appsforartists commented:
As long as we're cleaning stuff up, can you please change the license from MIT to GPL 3, to reflect changes made upstream?  I've been meaning to send one for that. 

Also, how do you decide when it's okay to use `with`?  I see one case in this PR where you're using it, and another where you broke one up into `lib.` calls.

GaetanLepage commented:
> As long as we're cleaning stuff up, can you please change the license from MIT to GPL 3, to reflect changes made upstream? I've been meaning to send one for that.

Sure, done.



> Also, how do you decide when it's okay to use `with`? I see one case in this PR where you're using it, and another where you've decomposed one into `lib.` calls.

It is generally recommended to avoid the `with` pattern as they can be quite unpredictable. I tend to remove them (especially from `meta`) when I come across them.

NickCao commented on pkgs/by-name/ha/handheld-daemon/package.nix:
```diff
@@ -1,35 +1,37 @@
   };
 
-  propagatedBuildInputs = with python3.pkgs; [
+  dependencies = with python3Packages; [
```
Please also set `build-system`.

GaetanLepage replied:
Sure, done.

NickCao commented on pkgs/by-name/ha/handheld-daemon/package.nix:
```diff
@@ -1,35 +1,41 @@
+  ];
+
+  dependencies = with python3Packages; [
     evdev
     hidapi
     kmod
+    lsof
+    pyserial
     pyyaml
     rich
     setuptools
     toybox
     xlib
```
The dependency list here kind of mismatches https://github.com/hhd-dev/hhd/blob/v3.6.2/pyproject.toml#L19-L24 ?

appsforartists commented:
Yeah, ActionScript/JavaScript had a similar period ~15y ago where they decided `with` was an antipattern because it made the scope more implicit. 

Thanks for the reply.

GaetanLepage commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 361364`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>handheld-daemon</li>
    <li>handheld-daemon.dist</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>handheld-daemon</li>
    <li>handheld-daemon.dist</li>
  </ul>
</details>

---
### `x86_64-darwin`

---
### `aarch64-darwin`

NickCao commented:
We should still patch (or wrap) these occurrences of subprocess.run: https://github.com/search?q=repo%3Ahhd-dev%2Fhhd%20subprocess&type=code

appsforartists reviewed (changes requested):
Since this package doesn't have tests, and is for hardware that you're presumably not testing on, I don't think removing dependencies is a good idea.  We didn't just arbitrarily add those.  You can search the hhd codebase and [see they're being used](https://github.com/search?q=repo%3Ahhd-dev%2Fhhd+lsof&type=code).

I don't want this package to break in the name of ostensible clean up.

toast003 commented on pkgs/by-name/ha/handheld-daemon/package.nix:
```diff
@@ -1,35 +1,35 @@
-    toybox
     xlib
-    pyserial
-    lsof
```
Please do not remove kmod and lsof from the dependencies, handheld-daemon uses them [here](https://github.com/hhd-dev/hhd/blob/c664e1530301a893648e6b09b6d1bc7b41568197/src/hhd/controller/physical/imu.py#L445) and [here](https://github.com/hhd-dev/hhd/blob/c664e1530301a893648e6b09b6d1bc7b41568197/src/hhd/plugins/overlay/x11.py#L136) respectively

appsforartists replied:
I also see:

- `toybox "id"`
- `systemd "udevadm"`
- `systemd "systemd-inhibit"`
- `dbus "dbus-monitor"`
- `efibootmgr`

You can see why I'm nervous this approach might break something.

GaetanLepage commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 361364`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>handheld-daemon</li>
    <li>handheld-daemon.dist</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>handheld-daemon</li>
    <li>handheld-daemon.dist</li>
  </ul>
</details>

---
### `x86_64-darwin`

---
### `aarch64-darwin`

toast003 commented on pkgs/by-name/ha/handheld-daemon/package.nix:
```diff
@@ -1,63 +1,96 @@
+        --replace-fail "lsof" "${lib.getExe lsof}"
+
+      substituteInPlace src/hhd/plugins/plugin.py \
+        --replace-fail "id" "${lib.getExe' toybox "id"}"
```
There's [quite a bit](https://github.com/search?q=repo%3Ahhd-dev%2Fhhd+path%3Aplugin.py+id&type=code) of variables that have `id` on their name so this breaks handheld-daemon.
```suggestion
        --replace-fail '"id"' '"${lib.getExe' toybox "id"}"'
```
This change would make it only change the path of `id`, which is what we want

GaetanLepage replied:
Good catch, done.

GaetanLepage commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 361364`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>handheld-daemon</li>
    <li>handheld-daemon.dist</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>handheld-daemon</li>
    <li>handheld-daemon.dist</li>
  </ul>
</details>

---
### `x86_64-darwin`

---
### `aarch64-darwin`

appsforartists commented on pkgs/by-name/ha/handheld-daemon/package.nix:
```diff
@@ -1,63 +1,96 @@
-  propagatedBuildInputs = with python3.pkgs; [
+  postPatch =
+    let
+      udevadm = lib.getExe' systemd "udevadm";
```
I know you're doing this to avoid repeating yourself, but I suspect being congruent with the other calls is more valuable than saving two lookups.

I suggest either:

- inline `udevadm` like the other dependencies, or
- make variables for all the dependencies like `chmodPath = lib.getExe' toybox "chmod"` and use them (`${ chmodPath }`) for all substitutions below.

GaetanLepage replied:
Yes, fair. I inlined everything.

appsforartists commented on pkgs/by-name/ha/handheld-daemon/package.nix:
```diff
@@ -1,63 +1,96 @@
+      # handheld-daemon contains a fork of the python module `hid`, so this hook
+      # is borrowed from the `hid` derivation.
+      hidapi=${lib.getLib hidapi}/lib/
+      test -d $hidapi || { echo "ERROR: $hidapi doesn't exist, please update/fix this build expression."; exit 1; }
```
This test and comment are holdovers from the `hid` derivation.  Since you've changed so much, you can probably remove them here (and might want to open a PR for that package with similar changes to what you've done here).

GaetanLepage replied:
Yes, you are right. I at least removed it from here.

appsforartists commented on pkgs/by-name/ha/handheld-daemon/package.nix:
```diff
@@ -1,63 +1,96 @@
+      substituteInPlace src/hhd/controller/lib/hid.py \
+        --replace-fail "libhidapi" "$hidapi/libhidapi"
+
+      # The generated udev rules point to /bin/chmod, which does not exist in NixOS
```
Maybe make this comment more generic and move it to the top of `postPatch`

GaetanLepage replied:
Done.

GaetanLepage commented:
I think that this PR is in a decent state now.

toast003 reviewed (approved):
Looks good to me, although hhd 3.7.0 came out already.
Should we update this PR to it or do it on a separate one?

GaetanLepage committed:
handheld-daemon: 3.6.2 -> 3.7.0

Diff: https://github.com/hhd-dev/hhd/compare/None...v3.7.0

Changelog: https://github.com/hhd-dev/hhd/releases/tag/v3.7.0

GaetanLepage commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 361364`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>handheld-daemon</li>
    <li>handheld-daemon.dist</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>handheld-daemon</li>
    <li>handheld-daemon.dist</li>
  </ul>
</details>

---
### `x86_64-darwin`

---
### `aarch64-darwin`

toast003 commented on pkgs/by-name/ha/handheld-daemon/package.nix:
```diff
@@ -1,63 +1,90 @@
+    substituteInPlace src/hhd/plugins/overlay/x11.py \
+      --replace-fail "lsof" "${lib.getExe lsof}"
+
+    substituteInPlace src/hhd/plugins/plugin.py \
+      --replace-fail '"id"' '"${lib.getExe' coreutils "id"}"'
+  '';
```
```suggestion
    substituteInPlace src/hhd/plugins/plugin.py \
      --replace-fail '"id"' '"${lib.getExe' coreutils "id"}"'
    substituteInPlace src/hhd/plugins/overlay/power.py \
      --replace-fail '"efibootmgr"' '"${lib.getExe' efibootmgr "id"}"'
  '';
```

[This file](https://github.com/hhd-dev/hhd/blob/c664e1530301a893648e6b09b6d1bc7b41568197/src/hhd/plugins/overlay/power.py) uses both efibootmgr and systemctl. Only suggesting efibootctl since systemctl seems to already be in `handheld-daemon`'s path

toast003 commented:
Please update the PR title and description whenever you can


Dec 9, 2024, 10:34 AM by ***@***.***:

>
>
>
> nixpkgs-review>  result
>
> Generated using > nixpkgs-review <https://github.com/Mic92/nixpkgs-review>> .
>
>
> Command: > nixpkgs-review pr 361364
>
>
>
> x86_64-linux
> ✅ 2 packages built:
> handheld-daemon
> handheld-daemon.dist
>
>
> aarch64-linux
> ✅ 2 packages built:
> handheld-daemon
> handheld-daemon.dist
>
>
> x86_64-darwin
>
>
> aarch64-darwin
>
>
> —
> Reply to this email directly, > view it on GitHub <https://github.com/NixOS/nixpkgs/pull/361364#issuecomment-2527398290>> , or > unsubscribe <https://github.com/notifications/unsubscribe-auth/AJJUMASMODNJZUEMBJXTDST2EVP3HAVCNFSM6AAAAABS5EVIBGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMRXGM4TQMRZGA>> .
> You are receiving this because you were mentioned.> Message ID: > <NixOS/nixpkgs/pull/361364/c2527398290> @> github> .> com>
>
>
>

toast003 commented:
Looks good to me

GaetanLepage commented:
> Looks good to me, we can have this merged now

Thank you for the review !
